### PR TITLE
Attempt to support realloc()

### PR DESCRIPTION
**Pull request by versokova**
I found in code `realloc(p,0);` used as free(). Predator crashs on `warning: ignoring return value of alloca/malloc/calloc/realloc`

```
int main() {
    malloc(4);
    return 0;
}
```

There is an allocated object on heap, but no return value for malloc(). There is a memory leak, but predator didn't find it. It can't work unless we use the return value of malloc(). So I added a condition to invalidate the built-in functions. Can be?

for malloc, it crash here 
```
cl/pointsto_fics.cc:221:        pair.caller.operand = &opList[0];
```

### CODE REVIEW

File: sl/symheap.cc [hunk 1]
```diff
@@ -2925,13 +2925,13 @@ void SymHeapCore::addNeq(TValId v1, TValId v2)
...
-        CL_DEBUG("SymHeap::neqOp() refuses to add an extraordinary Neq predicate");
-        return;
-    }
+//     const EValueTarget code1 = this->valTarget(v1);
+//     const EValueTarget code2 = this->valTarget(v2);
+//
+//     if (VT_UNKNOWN != code1 && VT_UNKNOWN != code2) {
+//         CL_DEBUG("SymHeap::neqOp() refuses to add an extraordinary Neq predicate");
+//         return;
+//     }
```
Comment (versokova) [inline sl/symheap.cc]: How can I correctly add an edge between two pointers? (problem with tests `test-041{0,3,7}.c`)

**Comment by kdudka** [comment]
Sorry for the delay!

You cannot (and should not need to add `Neq` predicates) between addresses of non-abstract objects.  A pair of addresses to valid separately allocated objects of kind `OK_REGION` are always seen as non-equal by `proveNeq()`.

So you can safely remove the call to `sh_.addNeq()` in `execHeapRealloc()`.  If it does not work, there must be a bug somewhere else.  I found the code of `execHeapRealloc()` difficult to follow because of the interleaved operations on `sh_` and `resizeHeap`.  I would suggest to split the code modelling the in-place reallocation and the move semantics to separate functions.

Note that I needed commit 9348b4bfb42d679f53d7bcb0a73c9df4c14648ad to be able to verify `test-0257` with `gcc-8.3.0`.

**Comment by kdudka** [comment]
As for the crash in `cl/pointsto_fics.cc`, could we just fix it like this?
```
--- a/cl/pointsto_fics.cc
+++ b/cl/pointsto_fics.cc
@@ -214,7 +214,9 @@ bool isKnownModel(const Insn *insn, TBindPairs &pairs)
         CL_BREAK_IF(opList.size() != 3);
         TBindPair pair;

-        CL_BREAK_IF(opList[0].code == CL_OPERAND_VOID);
+        if (opList[0].code == CL_OPERAND_VOID)
+            // return value of malloc() not used
+            return false;

         pair.code = BINDPAIR_HEAP;
         pair.callee.uid = -generateMallocUid(insn);
```

**Comment by versokova** [comment]
No need to apologize. Thank you for comments. I agree with splitting `sh_` and `resizeHeap`. (I do it later).

>  A pair of addresses to valid separately allocated objects of kind `OK_REGION` are always seen as non-equal by `proveNeq()`.

This is true, but not our case. After `p = realloc(q, size)`, `p` is not equal to `q`, but `q` is freed. So, I tried something in `comparePointers` [symproc.cc] like:
```
        //points to
        const TObjId obj1 = sh.objByAddr(root1);
        const TObjId obj2 = sh.objByAddr(root2);

        const EObjKind objK1 = sh.objKind(obj1);
        const EObjKind objK2 = sh.objKind(obj2);

        CmpOpTraits cTraits;
        if (!describeCmpOp(&cTraits, code)) {
            CL_WARN("binary operator not implemented yet");
            return sh.valCreate(VT_UNKNOWN, VO_UNKNOWN);
        }

        if (objK1 == OK_REGION && objK2 == OK_REGION && obj1!=obj2 && 
           (!sh.isValid(obj1) || !sh.isValid(obj2)) &&
            cTraits.preserveEq && cTraits.preserveNeq) {
            return boolToVal(cTraits.negative);
        }
```
But it is not safe, because you can reuse pointer in next allocation (`test-0238.c`). And you don't remember that `p` is allocated before you freed `q` without explicit neq edge. Or I missed something? Is there any way to safely add a neq edge?

**Comment by kdudka** [comment]
OK, got it.  Predator knows that properly allocated objects never overlap with each other but it fails to prove non-equality of addresses of invalid objects.  This kind of reasoning has never been supported by Predator and it is not specific to `realloc()`.  Consider the following example:

```
#include <stdlib.h>
#include <verifier-builtins.h>

int main() {
    void *const x = malloc(4);
    void *const y = malloc(4);
    if (x == y)
        ___sl_error("This is provably unreachable.");

    free(x);
    if (x == y)
        ___sl_error("This is not reachable but Predator thinks it is.");

    free(y);
    if (x == y)
        ___sl_error("This is not reachable but Predator thinks it is.");
}
```

If `(x != y)` was true initially, it must be true after executing `free(x)` and/or `free(y)`.  But Predator does not track relation of which objects coexisted at the same time to be able to prove non-equivalence of their addresses later on.

I am afraid that a proper fix would require some rethinking of the heap representation.  Do we have a real-world use case for verifying such programs?

**Comment by versokova** [comment]
That sounds acceptable. One of the SV-COMP benchmarks cannot be called real-world use case. I delete the condition from `test-0258.c` too.

**Comment by kdudka** [comment]
Merging.  Thanks for the contribution!

**Pull request closed without merging**